File: adapters/custom-engine/translator.go
package main

import (
	"github.com/usip/backend/schema"
)

// simWaypoint is the engine's internal waypoint representation.
type simWaypoint struct {
	lat, lon, alt float64
	speedMS       float64
	holdTimeS     uint32
}

// simEntity is the engine's mutable per-entity state during a run.
type simEntity struct {
	id    string
	name  string
	side  schema.Side
	etype schema.EntityType

	// Live kinematic state
	lat, lon, alt float64
	headingDeg    float64
	speedMS       float64
	alive         bool

	// Mission
	waypoints  []simWaypoint
	roe        schema.ROE
	wpIndex    int
	holdUntilS float64 // sim-time second until which the entity loiters

	// Derived capabilities (meters)
	sensorRangeM float64
	weaponRangeM float64

	// Bookkeeping: enemies already detected (avoid duplicate detection events)
	detected map[string]bool
}

// capability holds the default sensor/weapon ranges and base lethality for a
// given entity type. These are deliberately simple — this is a reference engine.
type capability struct {
	sensorRangeM float64
	weaponRangeM float64
	basePk       float64 // base probability of kill when it engages
}

// capabilityTable maps entity type to default capabilities. Engine-specific
// overrides can later come from Entity.attributes.
var capabilityTable = map[schema.EntityType]capability{
	schema.EntityType_ENTITY_TYPE_FIXED_WING:     {sensorRangeM: 80000, weaponRangeM: 40000, basePk: 0.7},
	schema.EntityType_ENTITY_TYPE_ROTARY_WING:    {sensorRangeM: 30000, weaponRangeM: 8000, basePk: 0.6},
	schema.EntityType_ENTITY_TYPE_UAV:            {sensorRangeM: 60000, weaponRangeM: 15000, basePk: 0.5},
	schema.EntityType_ENTITY_TYPE_GROUND_VEHICLE: {sensorRangeM: 8000, weaponRangeM: 4000, basePk: 0.6},
	schema.EntityType_ENTITY_TYPE_SURFACE_VESSEL: {sensorRangeM: 100000, weaponRangeM: 50000, basePk: 0.65},
	schema.EntityType_ENTITY_TYPE_SUBMARINE:      {sensorRangeM: 40000, weaponRangeM: 20000, basePk: 0.7},
	schema.EntityType_ENTITY_TYPE_RADAR_SENSOR:   {sensorRangeM: 150000, weaponRangeM: 0, basePk: 0},
	schema.EntityType_ENTITY_TYPE_BASE_FOB:       {sensorRangeM: 50000, weaponRangeM: 0, basePk: 0},
	schema.EntityType_ENTITY_TYPE_MISSILE:        {sensorRangeM: 20000, weaponRangeM: 10000, basePk: 0.85},
}

// defaultCapability is used for any type not in the table.
var defaultCapability = capability{sensorRangeM: 10000, weaponRangeM: 5000, basePk: 0.5}

func capabilityFor(t schema.EntityType) capability {
	if c, ok := capabilityTable[t]; ok {
		return c
	}
	return defaultCapability
}

// translate converts a canonical ScenarioProto into the engine's internal
// entity list, attaching each entity's mission waypoints/ROE.
func translate(s *schema.ScenarioProto) []*simEntity {
	missionByEntity := make(map[string]*schema.EntityMission, len(s.Missions))
	for _, m := range s.Missions {
		missionByEntity[m.EntityId] = m
	}

	out := make([]*simEntity, 0, len(s.Entities))
	for _, e := range s.Entities {
		cap := capabilityFor(e.Type)
		se := &simEntity{
			id:           e.Id,
			name:         e.Name,
			side:         e.Side,
			etype:        e.Type,
			alive:        true,
			roe:          schema.ROE_ROE_WEAPONS_TIGHT,
			sensorRangeM: cap.sensorRangeM,
			weaponRangeM: cap.weaponRangeM,
			detected:     make(map[string]bool),
		}
		if e.Position != nil {
			se.lat = e.Position.Lat
			se.lon = e.Position.Lon
			se.alt = e.Position.AltM
		}
		if m := missionByEntity[e.Id]; m != nil {
			se.roe = m.Roe
			for _, w := range m.Waypoints {
				se.waypoints = append(se.waypoints, simWaypoint{
					lat:       w.Lat,
					lon:       w.Lon,
					alt:       w.AltM,
					speedMS:   w.SpeedMs,
					holdTimeS: w.HoldTimeS,
				})
			}
		}
		out = append(out, se)
	}
	return out
}

File: adapters/panopticon/translator.py
"""Translate a USIP ScenarioProto into the engine's internal model.

The built-in placeholder engine consumes the SimEntity model below. The
`to_panopticon_scenario` function is the seam where a real Panopticon
integration would instead emit Panopticon's own scenario format.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List

from protostubs import entity_pb2, mission_pb2, scenario_pb2


@dataclass
class Capability:
    sensor_range_m: float
    weapon_range_m: float
    base_pk: float  # probability of kill per engagement


# Default sensor/weapon ranges and lethality by entity type. Mirrors the
# custom-engine's table so the two reference engines are roughly comparable.
CAPABILITY_TABLE: Dict[int, Capability] = {
    entity_pb2.ENTITY_TYPE_FIXED_WING: Capability(80000, 40000, 0.7),
    entity_pb2.ENTITY_TYPE_ROTARY_WING: Capability(30000, 8000, 0.6),
    entity_pb2.ENTITY_TYPE_UAV: Capability(60000, 15000, 0.5),
    entity_pb2.ENTITY_TYPE_GROUND_VEHICLE: Capability(8000, 4000, 0.6),
    entity_pb2.ENTITY_TYPE_SURFACE_VESSEL: Capability(100000, 50000, 0.65),
    entity_pb2.ENTITY_TYPE_SUBMARINE: Capability(40000, 20000, 0.7),
    entity_pb2.ENTITY_TYPE_RADAR_SENSOR: Capability(150000, 0, 0.0),
    entity_pb2.ENTITY_TYPE_BASE_FOB: Capability(50000, 0, 0.0),
    entity_pb2.ENTITY_TYPE_MISSILE: Capability(20000, 10000, 0.85),
}
DEFAULT_CAPABILITY = Capability(10000, 5000, 0.5)


def capability_for(etype: int) -> Capability:
    return CAPABILITY_TABLE.get(etype, DEFAULT_CAPABILITY)


@dataclass
class Waypoint:
    lat: float
    lon: float
    alt_m: float
    speed_ms: float
    hold_time_s: int


@dataclass
class SimEntity:
    id: str
    name: str
    side: int
    etype: int
    lat: float
    lon: float
    alt_m: float
    roe: int
    waypoints: List[Waypoint] = field(default_factory=list)
    # mutable run state
    alive: bool = True
    heading_deg: float = 0.0
    speed_ms: float = 0.0
    wp_index: int = 0
    hold_until_s: float = 0.0
    sensor_range_m: float = 0.0
    weapon_range_m: float = 0.0
    detected: Dict[str, bool] = field(default_factory=dict)


def translate(scenario: "scenario_pb2.ScenarioProto") -> List[SimEntity]:
    """ScenarioProto -> internal SimEntity list with missions attached."""
    mission_by_entity: Dict[str, "mission_pb2.EntityMission"] = {
        m.entity_id: m for m in scenario.missions
    }

    out: List[SimEntity] = []
    for e in scenario.entities:
        cap = capability_for(e.type)
        se = SimEntity(
            id=e.id,
            name=e.name,
            side=e.side,
            etype=e.type,
            lat=e.position.lat,
            lon=e.position.lon,
            alt_m=e.position.alt_m,
            roe=mission_pb2.ROE_WEAPONS_TIGHT,
            sensor_range_m=cap.sensor_range_m,
            weapon_range_m=cap.weapon_range_m,
        )
        m = mission_by_entity.get(e.id)
        if m is not None:
            se.roe = m.roe
            for w in m.waypoints:
                se.waypoints.append(
                    Waypoint(w.lat, w.lon, w.alt_m, w.speed_ms, w.hold_time_s)
                )
        out.append(se)
    return out


def to_panopticon_scenario(scenario: "scenario_pb2.ScenarioProto") -> dict:
    """SEAM: convert ScenarioProto to Panopticon's scenario format.

    Implement this when wiring the real Panopticon engine. Panopticon scenarios
    are JSON documents describing sides, aircraft/ships/facilities, and their
    routes/missions; map ScenarioProto entities + missions onto that schema and
    map our EntityType enum to Panopticon platform types.
    """
    raise NotImplementedError(
        "Panopticon scenario translation is not implemented yet. "
        "The built-in placeholder engine is used instead (see engine.py)."
    )
